packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmObjectHelper.kt

```diff
@@ -759,6 +761,141 @@ internal object RealmObjectHelper {
         }
     }
 
+    @Suppress("LongParameterList", "NestedBlockDepth", "LongMethod", "ComplexMethod")
+    internal fun assignValuesOnUnmanagedObject(
+        target: BaseRealmObject,
+        source: BaseRealmObject,
+        mediator: Mediator,
+        currentDepth: Int,
+        maxDepth: Int,
+        closeAfterCopy: Boolean,
+        cache: ManagedToUnmanagedObjectCache
+    ) {
+        val metadata: ClassMetadata = source.realmObjectReference!!.metadata
+        for (property in metadata.properties) {
+            val accessor = property.accessor
+                ?: sdkError("Typed object should always have an accessor")
+            if (property.isComputed) {
+                continue
+            }
+            accessor as KMutableProperty1<BaseRealmObject, Any?>
+            when (property.collectionType) {
+                CollectionType.RLM_COLLECTION_TYPE_NONE -> when (property.type) {
+                    PropertyType.RLM_PROPERTY_TYPE_OBJECT -> {
+                        if (currentDepth == maxDepth) {
+                            accessor.set(target, null)
+                        } else {
+                            val realmObject: BaseRealmObject? = accessor.get(source) as BaseRealmObject?
+                            if (realmObject != null) {
+                                accessor.set(
+                                    target,
+                                    createDetachedCopy(
```

Comment:
Can we leverage an implementation using `tailrec`?

Reply:
I don't think so, judging from the document, the tailrec function must be the last method called inside the function and this function is called from multiple places inside here and from the public `copyFromRealm` method.

But neat idea actually ... I didn't really have `tailrec` on my radar.